x-pack/plugin/ent-search/src/main/java/org/elasticsearch/xpack/application/connector/ConnectorStateMachine.java

```diff
@@ -39,6 +39,16 @@ public class ConnectorStateMachine {
      * @param next The proposed next state of the connector.
      */
     public static boolean isValidTransition(ConnectorStatus current, ConnectorStatus next) {
-        return VALID_TRANSITIONS.getOrDefault(current, Collections.emptySet()).contains(next);
+        return validNextStates(current).contains(next);
+    }
+
+    public static void assertValidStateTransition(ConnectorStatus current, ConnectorStatus next)
+        throws ConnectorInvalidStatusTransitionException {
+        if (isValidTransition(current, next)) return;
+        throw new ConnectorInvalidStatusTransitionException(current, next);
+    }
+
+    public static Set<ConnectorStatus> validNextStates(ConnectorStatus current) {
+        return VALID_TRANSITIONS.getOrDefault(current, Collections.emptySet());
```

Comment:
Question (somewhat unrelated to this exact PR): why is `error` not a valid state following `needs_configuration`?

Reply:
I couldn't find an exact reference in the current logic, but I agree that it makes sense. I added this as a valid transition.